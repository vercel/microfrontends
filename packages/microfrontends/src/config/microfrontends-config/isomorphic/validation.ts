import { pathToRegexp, parse as parsePathRegexp } from 'path-to-regexp';
import type {
  ApplicationId,
  PathGroup,
  ApplicationRouting,
  ChildApplication as ChildApplicationConfig,
} from '../../schema/types';
import { MicrofrontendError } from '../../errors';
import { isDefaultApp } from '../../schema/utils/is-default-app';

const LIST_FORMATTER = new Intl.ListFormat('en', {
  style: 'long',
  type: 'conjunction',
});

const VALID_ASSET_PREFIX_REGEXP = /^[a-z](?:[a-z0-9-]*[a-z0-9])?$/;

/**
 * Validate all paths in a configuration - ensures paths do not overlap
 */
export const validateConfigPaths = (
  applicationConfigsById?: ApplicationRouting,
): void => {
  if (!applicationConfigsById) {
    return;
  }

  const pathsByApplicationId = new Map<
    PathGroup['paths'][number],
    {
      applications: ApplicationId[];
      matcher: RegExp;
      applicationId?: ApplicationId;
    }
  >();
  const errors: string[] = [];

  for (const [id, app] of Object.entries(applicationConfigsById)) {
    if (isDefaultApp(app)) {
      // default applications do not have routing
      continue;
    }

    for (const pathMatch of app.routing) {
      for (const path of pathMatch.paths) {
        const maybeError = validatePathExpression(path);
        if (maybeError) {
          errors.push(maybeError);
        } else {
          const existing = pathsByApplicationId.get(path);
          if (existing) {
            existing.applications.push(id);
          } else {
            pathsByApplicationId.set(path, {
              applications: [id],
              matcher: pathToRegexp(path),
              applicationId: id,
            });
          }
        }
      }
    }
  }
  const entries = Array.from(pathsByApplicationId.entries());

  for (const [path, { applications: ids, matcher, applicationId }] of entries) {
    if (ids.length > 1) {
      errors.push(
        `Duplicate path "${path}" for applications "${ids.join(', ')}"`,
      );
    }

    for (const [
      matchPath,
      { applications: matchIds, applicationId: matchApplicationId },
    ] of entries) {
      if (path === matchPath) {
        // we're comparing to ourselves, so skip
        continue;
      }

      if (applicationId === matchApplicationId) {
        // we're comparing to paths within our own application, which are allowed to overlap, so skip
        continue;
      }

      if (matcher.test(matchPath)) {
        const source = `"${path}" of application${ids.length > 0 ? 's' : ''} ${ids.join(', ')}`;
        const destination = `"${matchPath}" of application${matchIds.length > 0 ? 's' : ''} ${matchIds.join(', ')}`;

        errors.push(
          `Overlapping path detected between ${source} and ${destination}`,
        );
      }
    }
  }

  if (errors.length) {
    throw new MicrofrontendError(
      `Invalid paths: ${errors.join(', ')}. See supported paths in the documentation https://vercel.com/docs/microfrontends/path-routing#supported-path-expressions.`,
      {
        type: 'config',
        subtype: 'conflicting_paths',
      },
    );
  }
};

// From https://github.com/pillarjs/path-to-regexp/blob/b0778f5e8e6c6e9ee4e2f5b34e877cc5229f8036/src/index.ts#L143
const PATH_DEFAULT_PATTERN = '[^\\/#\\?]+?';

function validatePathExpression(path: string): string | undefined {
  try {
    const tokens = parsePathRegexp(path);
    if (/(?<!\\)\{/.test(path)) {
      return `Optional paths are not supported: ${path}`;
    }
    if (/(?<!\\|\()\?/.test(path)) {
      return `Optional paths are not supported: ${path}`;
    }
    if (/\/[^/]*(?<!\\):[^/]*(?<!\\):[^/]*/.test(path)) {
      return `Only one wildcard is allowed per path segment: ${path}`;
    }
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      if (token === undefined) {
        return `token ${i} in ${path} is undefined, this shouldn't happen`;
      }
      if (typeof token !== 'string') {
        if (!token.name) {
          return `Only named wildcards are allowed: ${path} (hint: add ":path" to the wildcard)`;
        }
        if (
          token.pattern !== PATH_DEFAULT_PATTERN &&
          // Allows (a|b|c) and ((?!a|b|c).*) regex
          // Only limited regex is supported for now, due to performance considerations
          // Allows all letters, numbers, and hyphens. Other characters must be escaped.
          !/^(?<allowed>[\w-~]+(?:\|[^:|()]+)+)$|^\(\?!(?<disallowed>[\w-~]+(?:\|[^:|()]+)*)\)\.\*$/.test(
            token.pattern.replace(/\\./g, ''),
          )
        ) {
          return `Path ${path} cannot use unsupported regular expression wildcard. If the path includes special characters, they must be escaped with backslash (e.g. '\\(')`;
        }
        if (token.modifier && i !== tokens.length - 1) {
          return `Modifier ${token.modifier} is not allowed on wildcard :${token.name} in ${path}. Modifiers are only allowed in the last path component`;
        }
      }
    }
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    return `Path ${path} could not be parsed into regexp: ${message}`;
  }
  return undefined;
}

/**
 * Validate all paths in an application - ensures paths are the correct format
 */
export const validateAppPaths = (
  name: string,
  app: ChildApplicationConfig,
): void => {
  // validate routes
  for (const group of app.routing) {
    for (const p of group.paths) {
      if (p === '/') {
        continue;
      }
      if (p.endsWith('/')) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must not end with a slash.`,
          { type: 'application', subtype: 'invalid_path' },
        );
      }

      if (!p.startsWith('/')) {
        throw new MicrofrontendError(
          `Invalid path for application "${name}". ${p} must start with a slash.`,
          { type: 'application', subtype: 'invalid_path' },
        );
      }
    }
  }
  if (app.assetPrefix) {
    if (!VALID_ASSET_PREFIX_REGEXP.test(app.assetPrefix)) {
      throw new MicrofrontendError(
        `Invalid asset prefix for application "${name}". ${app.assetPrefix} must start with a lowercase letter and contain only lowercase letters, numbers, and hyphens.`,
        { type: 'application', subtype: 'invalid_asset_prefix' },
      );
    }
    if (
      app.assetPrefix !== `vc-ap-${name}` &&
      !app.routing.some(
        (group) =>
          group.paths.includes(`/${app.assetPrefix}/:path*`) && !group.flag,
      )
    ) {
      throw new MicrofrontendError(
        `When \`assetPrefix\` is specified, \`/${app.assetPrefix}/:path*\` must be added the routing paths for the application. Changing the asset prefix is not a forwards and backwards compatible change, and the custom asset prefix should be added to \`paths\` and deployed before setting the \`assetPrefix\` field.`,
        { type: 'application', subtype: 'invalid_asset_prefix' },
      );
    }
  }
};

/**
 * Make sure only one `Application` defines routing
 * */
export const validateConfigDefaultApplication = (
  applicationConfigsById?: ApplicationRouting,
): void => {
  if (!applicationConfigsById) {
    return;
  }

  const applicationsWithoutRouting = Object.entries(
    applicationConfigsById,
  ).filter(([, app]) => isDefaultApp(app));
  const numApplicationsWithoutRouting = applicationsWithoutRouting.reduce(
    (acc) => {
      return acc + 1;
    },
    0,
  );

  if (numApplicationsWithoutRouting === 0) {
    throw new MicrofrontendError(
      'No default application found. At least one application needs to be the default by omitting routing.',
      { type: 'config', subtype: 'no_default_application' },
    );
  }

  if (numApplicationsWithoutRouting > 1) {
    const applicationNamesMissingRouting = applicationsWithoutRouting.map(
      ([name]) => name,
    );
    throw new MicrofrontendError(
      `All applications except for the default app must contain the "routing" field. Applications that are missing routing: ${LIST_FORMATTER.format(applicationNamesMissingRouting)}.`,
      { type: 'config', subtype: 'multiple_default_applications' },
    );
  }
};
