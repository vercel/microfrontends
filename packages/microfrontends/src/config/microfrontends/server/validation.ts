import { parse } from 'jsonc-parser';
import { Ajv, type ErrorObject } from 'ajv';
import type { Config } from '../../schema/types';
import { MicrofrontendError } from '../../errors';
import { SCHEMA } from '../../schema/utils/load';

const LIST_FORMATTER = new Intl.ListFormat('en', {
  style: 'long',
  type: 'disjunction',
});

function formatAjvErrors(errors: ErrorObject[] | null | undefined): string[] {
  if (!errors) {
    return [];
  }
  const errorMessages: string[] = [];
  for (const error of errors) {
    // Ignore root errors that arise because AJV can't infer whether the config is a main config or a child config.
    // These errors are confusing for the user.
    if (
      error.instancePath === '' &&
      (error.keyword === 'anyOf' ||
        (error.keyword === 'required' &&
          error.params.missingProperty === 'partOf'))
    ) {
      continue;
    }

    // Ignore the first leading slash in the instance path to make the path more readable.
    const instancePath = error.instancePath.slice(1);
    const formattedInstancePath =
      instancePath === '' ? 'at the root' : `in field ${instancePath}`;

    // AJV errors are really cryptic, so this logic transforms them to nicer ones.
    // It is based on heuristics from introducing errors and observing what AJV produces.
    if (
      error.keyword === 'required' &&
      error.params.missingProperty === 'routing' &&
      instancePath.split('/').length === 2
    ) {
      errorMessages.push(
        `Unable to infer if ${instancePath} is the default app or a child app. This usually means that there is another error in the configuration.`,
      );
    } else if (
      error.keyword === 'anyOf' &&
      instancePath.split('/').length > 2
    ) {
      const anyOfErrors = errors.filter(
        (e) => e.instancePath === error.instancePath && e.keyword !== 'anyOf',
      );
      if (anyOfErrors.every((e) => e.keyword === 'type')) {
        const allowedTypes = LIST_FORMATTER.format(
          anyOfErrors.map((e) => {
            return e.keyword === 'type' ? String(e.params.type) : 'unknown';
          }),
        );
        errorMessages.push(
          `Incorrect type for ${instancePath}. Must be one of ${allowedTypes}`,
        );
      } else {
        errorMessages.push(
          `Invalid field for ${instancePath}. Possible error messages are ${LIST_FORMATTER.format(anyOfErrors.map((e) => e.message ?? ''))}`,
        );
      }
    } else if (
      error.keyword === 'additionalProperties' &&
      !(
        error.params.additionalProperty === 'routing' &&
        instancePath.split('/').length === 2
      )
    ) {
      errorMessages.push(
        `Property '${error.params.additionalProperty}' is not allowed ${formattedInstancePath}`,
      );
    } else if (error.keyword === 'required') {
      errorMessages.push(
        `Property '${error.params.missingProperty}' is required ${formattedInstancePath}`,
      );
    }
  }
  return errorMessages;
}

export function validateSchema(configString: string): Config {
  const parsedConfig = parse(configString) as Config;

  const ajv = new Ajv({ allowUnionTypes: true });
  const validate = ajv.compile(SCHEMA);
  const isValid = validate(parsedConfig);
  if (!isValid) {
    throw new MicrofrontendError(
      `Invalid microfrontends config:${formatAjvErrors(validate.errors)
        .map((error) => `\n - ${error}`)
        .join(
          '',
        )}\n\nSee https://openapi.vercel.sh/microfrontends.json for the schema.`,
      { type: 'config', subtype: 'does_not_match_schema' },
    );
  }

  return parsedConfig;
}
