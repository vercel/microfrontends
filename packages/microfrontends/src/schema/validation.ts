export { validateSchema } from '../config/microfrontends/server/validation';
