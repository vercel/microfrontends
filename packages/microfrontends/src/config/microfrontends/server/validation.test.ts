import { validateSchema } from './validation';

describe('validateSchema', () => {
  describe('valid configs do not throw errors', () => {
    it('only default app', () => {
      const config = {
        applications: {
          app1: {
            development: {
              local: 'localhost:3000',
              fallback: 'prod.com',
            },
          },
        },
      };
      expect(validateSchema(JSON.stringify(config))).toEqual(config);
    });

    it('full config', () => {
      const config = {
        $schema: 'https://openapi.vercel.sh/microfrontends.json',
        applications: {
          'nextjs-app-marketing': {
            development: {
              local: 3000,
              fallback: 'nextjs-app-marketing.vercel.app',
            },
          },
          'nextjs-app-docs': {
            development: {
              local: 3001,
            },
            routing: [
              {
                group: 'docs',
                paths: ['/docs', '/docs/:path*'],
              },
              {
                group: 'flagged-docs-paths',
                flag: 'flagged-docs-enabled',
                paths: ['/flagged/docs'],
              },
              {
                group: 'mid-path-wildcard',
                paths: ['/some-:hash.js'],
              },
              {
                group: 'excluding regex',
                paths: ['/city/:path((?!berlin|amsterdam).*)'],
              },
            ],
          },
        },
        options: {
          disableOverrides: true,
          localProxyPort: 3025,
        },
      };
      expect(validateSchema(JSON.stringify(config))).toEqual(config);
    });
  });

  describe('invalid configs throw errors', () => {
    it('additional root field', () => {
      const config = {
        applications: {
          app1: {
            development: {
              local: 'localhost:3000',
            },
          },
        },
        unnecessaryField: 'test',
      };
      expect(() =>
        validateSchema(JSON.stringify(config)),
      ).toThrowErrorMatchingSnapshot();
    });

    it('missing applications', () => {
      const config = {
        version: '1',
      };
      expect(() =>
        validateSchema(JSON.stringify(config)),
      ).toThrowErrorMatchingSnapshot();
    });

    it('invalid routing', () => {
      const config = {
        applications: {
          app1: {
            development: {
              local: 'localhost:3000',
            },
            routing: [
              {
                paths: ['/*'],
                flagg: 'invalid-flag',
              },
            ],
          },
        },
      };
      expect(() =>
        validateSchema(JSON.stringify(config)),
      ).toThrowErrorMatchingSnapshot();
    });

    it('invalid local field', () => {
      const config = {
        applications: {
          app1: {
            development: {
              local: false,
            },
          },
        },
      };
      expect(() =>
        validateSchema(JSON.stringify(config)),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
