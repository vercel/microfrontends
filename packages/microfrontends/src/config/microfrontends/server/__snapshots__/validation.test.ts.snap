// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateSchema invalid configs throw errors additional root field 1`] = `
"Invalid microfrontends config:
 - Property 'unnecessaryField' is not allowed at the root

See https://openapi.vercel.sh/microfrontends.json for the schema."
`;

exports[`validateSchema invalid configs throw errors invalid local field 1`] = `
"Invalid microfrontends config:
 - Property 'fallback' is required in field applications/app1/development
 - Unable to infer if applications/app1 is the default app or a child app. This usually means that there is another error in the configuration.

See https://openapi.vercel.sh/microfrontends.json for the schema."
`;

exports[`validateSchema invalid configs throw errors invalid routing 1`] = `
"Invalid microfrontends config:
 - Property 'flagg' is not allowed in field applications/app1/routing/0

See https://openapi.vercel.sh/microfrontends.json for the schema."
`;

exports[`validateSchema invalid configs throw errors missing applications 1`] = `
"Invalid microfrontends config:
 - Property 'applications' is required at the root

See https://openapi.vercel.sh/microfrontends.json for the schema."
`;
